### PR TITLE
Update README's guide on how to link with RNPM in the postinstall hook

### PR DESCRIPTION
**Pull request by geof90**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -166,11 +166,15 @@ Yes, in fact - it has nothing to do with it. What it does is just linking static
...
-  "postinstall": "rnpm link"
+"dependencies": {
+  "rnpm": "^1.4.1"
 },
+"scripts": { 
+  "postinstall": "cd ../.. && rnpm link <name>"
+}
 ...
 ```
+Note that the `cd ../..` in the postinstall script is needed to go up to the working directory of the parent project that is to be linked with your plugin.
```
Comment (grabbou) [inline README.md]: In 99% cases there's only one package.json and rnpm works just fine from the root folder. Are you having a different use-case?

Comment (Kureev) [inline README.md]: ?

Comment (geof90) [inline README.md]: Maybe I'm missing something. Usually users install our plugin from the root folder of their project, e.g npm install --save react-native-code-push. That will get saved in `./node_modules/react-native-code-push`, therefore, two directories deep from the parent project. The goal of this postinstall script is to link the plugin with the parent project, therefore doesn't it have to `cd` up two directories? I asked this in Discord but didnt really get a reply on this.

Comment (geof90) [inline README.md]: @grabbou @Kureev thoughts?

Comment (Kureev) [inline README.md]: I remember this. My question is addressed to @grabbou :)

**Comment by grabbou** [comment]
Ah, @geof90 but keep in mind postinstall hook is only intended for your project, not for the dependency, so there's no need to `../../` if I get that right.

**Comment by Kureev** [comment]
I think he wants to run an `rnpm link` on plugin install

**Comment by Kureev** [comment]
so the default dir would be `xxx/node_modules/rnpm-plugin-code-push`

**Comment by geof90** [comment]
@grabbou, I thought the whole point of putting `rnpm link` inside a `postinstall` hook is so that when a React Native developer wants to add a plugin to his/her app, he/her simply needs to `npm install` that plugin and it will be automatically linked.

In this case, my project (i.e `react-native-code-push`) IS the plugin, and therefore IS the dependency of another project that wants to use it as a plugin. I thought this was the whole point of using `postinstall`. Perhaps I'm mistaken.

**Comment by grabbou** [comment]
Ah, fair enough. I get the use case for now.

I still believe that postInstall hooks should stay on user project level only so that they have full control over what manipulates their projects (and whether they want to use rnpm or not). All in all, postInstall hook will run after every single npm install, so there's no difference for end users whether it's defined on their level, or on dependency level.

If you still want to do it, I'd see that as an extra readme section, although as I said - I am still not sure if that's the behaviour we want to promote / expose.

What do you think?

Also, re: your snippet, in order to avoid `rnpm` being fired too many times during npm install, I think developers should use the below postinstall hook instead:
```
cd ../../ && rnpm install package-name
```
so it does not scan the entire tree but goes directly to the installed project.

**Comment by geof90** [comment]
Are you perhaps saying that `rnpm` + `postinstall` is meant for the package.json of the RN project itself? Because if so, I fail to see how that is a practical scenario. Nobody would want to add that to their `package.json` dependency. Even in the scenario where multiple people are working on a project, the gradle xcodeproj files would be checked into their code repositories, therefore there's no need to add the `rnpm link` step to `postinstall` since the dependencies would already be linked. I always thought that my scenario (letting plugin developers simplify plugin linking by adding `rnpm link` to their `postinstall` hooks) is the main scenario this solution was trying to tackle.

**Comment by grabbou** [comment]
> Are you perhaps saying that rnpm + postinstall is meant for the package.json of the RN project itself? 

Yes.

> Even in the scenario where multiple people are working on a project, the gradle/xcodeproj files would be checked into their code repositories, therefore there's no need to add the rnpm link step to postinstall since the dependencies would already be linked. 

Well, that way you don't need package.json as long as you commit node_modules folder? To me that's quite reasonable to automatically perform some extra actions after dependencies are installed (similar to what people do with gulp build tasks etc.). 

>  I always thought that my scenario (letting plugin developers simplify plugin linking by adding rnpm link to their postinstall hooks) is the main scenario this solution was trying to tackle.

That's very interesting possibility, although I'd prefer avoiding rnpm link from being run that way. It not only makes the development trickier (clone repo, npm install - rnpm tries to link to ../../ which can be anything) but also produces a chaos in what's going on when you run npm install. Having a single point of entry makes it easy to control rnpm version for the user and to decide whether he wants automatic linking, no linking at all or anything.

**Comment by geof90** [comment]
Deleted my last comment. I see what you are saying. I had no idea that the `postinstall` would also run when you install dependencies, like what you are suggesting. So it seems your advice is, we should prescribe to our users to add `rnpm link` directly to their own package.json configs so that the automated linking will happen when they install our plugin as a dependency? If that works, awesome! And I guess I learnt something new about NPM today :p

I'll close this PR for now. Thanks for the reviews and help!

**Comment by grabbou** [comment]
Actually sorry - turns out I was wrong - postinstall won't run on the parent project when you explicitly specify dependency name to install. Leaving that opened then, because the issue you described was actually valid - in order for automatic linking to work, package authors have to add postinstall hooks in their package.jsons.

**Comment by geof90** [comment]
Ah... ok. Thats a bummer.

Actually I suggested to @Kureev on Discord about the possibility of you guys creating a simple script that does the `cd ../../` and calls `rnpm link` for us, so we could point the `postinstall` to that script instead. Like maybe `rnpm link-self`. That would be the ideal solution. In the absence of that, this workaround would have to do, unless you have better ideas!

**Comment by geof90** [comment]
Another solution could be adding logic in the `rnpm link` to detect if it is in the node_modules folder. Something like:
```
if (/node_modules\/plugin-name/.test(process.cwd())) {
  process.chdir('../..');
}
```

**Comment by geof90** [comment]
@grabbou just saw your comment edit. Your solution definitely works too. The only tradeoff is that it forces the user to install rnpm globally. Which is okay I guess.

**Comment by Kureev** [comment]
Well, I think `rnpm` **should** be installed globally. So I'm also up for @grabbou's variant. Make an install plugin and aggregate `npm install <package>` + `rnpm link <package>`. I'll take care about it.
ETA: this weekend

**Comment by grabbou** [comment]
@Kureev I think it should be rnpm install <package> to avoid breaking changes and make it super light-weight.

**Comment by geof90** [comment]
In this case, is there any value in attaching RNPM to postinstall hooks at all then? Seems like the answer going forward is no

**Comment by Kureev** [comment]
Yep, for me making something like `cd ../../ && rnpm link <package>` is a bit tricky. I think `rnpm install` will do the trick without any additional efforts from the plugin authors.

**Comment by grabbou** [comment]
+ @Kureev I'll add a warning when one calls `rnpm link <abc>` and we cannot locate `abc` that `oops, you have probably wanted to npm install it first. Do it or use rnpm install`

**Comment by Kureev** [comment]
Agree

**Comment by Kureev** [comment]
Anyway, I'm going to close this PR. Hope everybody agree on it.

**Comment by Kureev** [comment]
see https://github.com/rnpm/rnpm/pull/87

**Pull request closed without merging**